File: cpp/111. Minimum Depth of Binary Tree.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// DFS, Time: O(n), Space: O(n)
public:
    int minDepth(TreeNode* root) {
        if(!root) return 0;
        if(!root->left && !root->right) return 1;
        if(!root->left) return 1 + minDepth(root->right);
        if(!root->right) return 1 + minDepth(root->left);
        return 1 + min(minDepth(root->left), minDepth(root->right));
    }
};

class Solution {
// BFS, Time: O(n), Space: O(n)
public:
    int minDepth(TreeNode* root) {
        if(!root) return 0;
        queue<TreeNode*> q; q.push(root);
        int ans = 0;
        bool is_terminal = false;
        while(!q.empty() && !is_terminal) {
            int sz = q.size();
            ans += 1;
            for(int i = 0; i < sz; i++) {
                TreeNode *node = q.front(); q.pop();
                if(!node->left && !node->right) return ans;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
        }
        return -1;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/111-Minimum Depth of Binary Tree.py
from bintree_helper import *

class Solution:
    def minDepth(self, root) -> int:
        "DFS, Time: O(n), Space: O(n)"
        def dfs(node):
            if node is None: return 0
            l = dfs(node.left)
            r = dfs(node.right)
            if node.left and node.right: return min(l, r) + 1
            if node.left is None and node.right is None: return 1
            return max(l, r) + 1
        return dfs(root)
    def minDepth(self, root) -> int:
        "BFS, Time: O(n), Space: O(n)"
        from collections import deque
        if root is None: return 0
        q = deque([root])
        level = 1
        while q:
            n = len(q)
            for _ in range(n):
                node = q.popleft()
                if node.left is None and node.right is None:
                    return level
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
            level += 1
        return -1

root = deserialize("[3,9,20,null,null,15,7]")
result = Solution().minDepth(root)
assert result == 2

root = deserialize("[2,null,3,null,4,null,5,null,6]")
result = Solution().minDepth(root)
assert result == 5